tsl/src/chunk.c

```diff
@@ -1180,3 +1186,538 @@ chunk_merge_chunks(PG_FUNCTION_ARGS)
+
+		/*
+		 * Check that the split_at value actually produces a valid split. Note
+		 * that range_start is inclusive while range_end is non-inclusive. The
+		 * split_at value needs to produce partition ranges of at least length
+		 * 1.
+		 */
+		if (split_point < (slice->fd.range_start + 1) || split_point > (slice->fd.range_end - 2))
+		{
+			Oid outfuncid = InvalidOid;
+			bool isvarlena = false;
+
+			getTypeOutputInfo(splitdim_type, &outfuncid, &isvarlena);
+			Datum split_at_str = OidFunctionCall1(outfuncid, split_at);
+
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("cannot split chunk at %s", DatumGetCString(split_at_str))));
+		}
+	}
+	else
+		split_point = slice->fd.range_start + (interval_range / 2);
+
+	int64 old_end = slice->fd.range_end;
+
+	/* Update the slice range for the existing chunk */
+	slice->fd.range_end = split_point;
+	chunk_update_constraints(chunk, new_cube);
+
+	/* Update the slice for the new chunk */
+	slice->fd.range_start = split_point;
+	slice->fd.range_end = old_end;
+	slice->fd.id = 0; /* Must set to 0 to mark as new for it to be created */
+
+	/* Make updated constraints visible */
+	CommandCounterIncrement();
+	bool created = false;
+	Chunk *new_chunk = ts_chunk_find_or_create_without_cuts(ht,
+															new_cube,
+															NameStr(chunk->fd.schema_name),
+															NULL,
+															InvalidOid,
+															&created);
+	Ensure(created, "could not create chunk for split");
+	Assert(new_chunk);
+
+	int ti_options = TABLE_INSERT_SKIP_FSM;
+	TupleTableSlot *srcslot;
+	TableScanDesc scan;
+	MemoryContext oldcxt;
+	EState *estate;
+	ExprContext *econtext;
+	RelationSplitInfo *splitinfos[2];
+	struct VacuumCutoffs cutoffs;
+	char relpersistence = splitrel->rd_rel->relpersistence;
+
+	compute_rel_vacuum_cutoffs(splitrel, &cutoffs);
+
+	Oid new_heap_relid = make_new_heap(RelationGetRelid(splitrel),
+									   splitrel->rd_rel->reltablespace,
+									   splitrel->rd_rel->relam,
+									   relpersistence,
+									   AccessExclusiveLock);
+
+	Relation splitrel_new = table_open(new_heap_relid, AccessExclusiveLock);
+	Relation new_chunkrel = table_open(new_chunk->table_id, AccessExclusiveLock);
+
+	estate = CreateExecutorState();
+
+	/* Create the tuple slot */
+	srcslot = MakeSingleTupleTableSlot(RelationGetDescr(splitrel), table_slot_callbacks(splitrel));
+
+	/* Setup the state we need for each new chunk partition */
+	splitinfos[0] = relation_split_info_create(splitrel, splitrel_new, &cutoffs);
+	splitinfos[1] = relation_split_info_create(splitrel, new_chunkrel, &cutoffs);
+
+	/* Scan through the rows using SnapshotAny to see everything. */
+	scan = table_beginscan(splitrel, SnapshotAny, 0, NULL);
+
+	/*
+	 * Switch to per-tuple memory context and reset it for each tuple
+	 * produced, so we don't leak memory.
+	 */
+	econtext = GetPerTupleExprContext(estate);
+	oldcxt = MemoryContextSwitchTo(GetPerTupleMemoryContext(estate));
```

Comment:
```suggestion
	econtext = GetPerTupleExprContext(estate);
	oldcxt = MemoryContextSwitchTo(econtext);
```

Reply:
You suggestion isn't correct. `econtext` is an expr context and not a memory context.

Reply:
Sorry, my bad.